File: foundation/foundation-metadata/foundation-metadata-core/src/main/java/com/boozallen/aiops/core/metadata/MetadataAPI.java
package com.boozallen.aiops.core.metadata;

/*-
 * #%L
 * AIOps Core::Metadata::AIOps Metadata
 * %%
 * Copyright (C) 2021 Booz Allen
 * %%
 * This software package is licensed under the Booz Allen Public License. All Rights Reserved.
 * #L%
 */

import java.util.List;
import java.util.Map;

public interface MetadataAPI {
    /**
     * Create metadata.
     * @param metadata the metadata to create.
     */
    void createMetadata(MetadataModel metadata);

    /**
     * Get metadata from search criteria.
     * @param searchParams the list of search parameters by key-value pair.
     * @return a list of metadata matching the search parameters.
     */
    List<MetadataModel> getMetadata(Map<String, Object> searchParams);
}

File: foundation/aissemble-foundation-core-python/src/aissemble_core_metadata/metadata_api.py
###
# #%L
# aiSSEMBLE Foundation::aiSSEMBLE Core (Python)
# %%
# Copyright (C) 2021 Booz Allen
# %%
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# #L%
###
from abc import ABC, abstractmethod
from typing import Dict, List
from .metadata_model import MetadataModel


class MetadataAPI(ABC):
    """
    API for a metadata service.
    """

    @abstractmethod
    def create_metadata(self, metadata: MetadataModel) -> None:
        """
        Method to create metadata.
        """
        pass

    @abstractmethod
    def get_metadata(self, search_params: Dict[str, any]) -> List[MetadataModel]:
        """
        Method to get metadata from search criteria.
        """
        pass
